include/hip/hcc_detail/ocml_math_fwd.h

```diff
@@ -0,0 +1,431 @@
+/*
+Copyright (c) 2015 - present Advanced Micro Devices, Inc. All rights reserved.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#pragma once
+
+#if defined(__cplusplus)
+    extern "C" {
+#endif
+
+// BEGIN FLOAT
+__attribute__((const))
```

Comment:
CUDA/HIP does not infer __device__ attribute. device function can only call device function, therefore, these functions need __device__ attribute to work with hip-clang.

Reply:
I was unaware that this is now a shared repository, and that we're doing shared development for the two nowadays...

Reply:
Since the device code for HCC and hip-clang are similar, so hip-clang now uses large portion of code in hcc_details to save time. This works pretty fine so far.

Reply:
Yes but you have no dedicated unit tests, and create extra overhead for people that may not want to be involved with that project. The present situation outlines both of those aspects - since there are no hip-clang unit tests in this repository any breakage trickles down via the grapevine, inefficiently.

Reply:
hip-clang is not ready for unit test yet, but will be.

If it causes any regressions in hip-hcc it will be caught by existing unit test. And we will fix hip-clang issues.

Reply:
Actually hip-clang uses the same directed unit-tests as hip-hcc. Just need to set some environment variables. It is OK if you would not like to make changes that are needed by hip-clang. Aaron or I can add it in a separate commit.